File: 01_GOF23/DP_examples/Python/01_Creational/Prototype/Product.py
from abc import ABC, abstractmethod

class Color:
    Unknown = 0
    Red = 1
    Green = 2
    Blue = 3
    Yellow = 4

    @staticmethod
    def to_string(color):
        color_map = {
            Color.Red: "Red",
            Color.Green: "Green",
            Color.Blue: "Blue",
            Color.Yellow: "Yellow"
        }
        return color_map.get(color, "Unknown")

class IShape(ABC):
    @abstractmethod
    def name(self):
        pass

    @abstractmethod
    def draw(self):
        pass

    @abstractmethod
    def fill(self, color):
        pass

    @abstractmethod
    def clone(self):
        pass

class Shape(IShape):
    def __init__(self, id=0, name=""):
        self._id = id
        self._name = name

    def name(self):
        return self._name

    def draw(self):
        pass

    def fill(self, color):
        pass

    def clone(self):
        pass

class ShapeKind:
    Unknown = 0
    Circle = 1
    Square = 2
    Rectangle = 3

class Circle(Shape):
    def __init__(self, id=ShapeKind.Circle, name="Circle"):
        super().__init__(id, name)

    def draw(self):
        print("Draw a Circle")

    def fill(self, color):
        print(f"Fill with {Color.to_string(color)}")

    def clone(self):
        return Circle(self._id, self._name)

class Square(Shape):
    def __init__(self, id=ShapeKind.Square, name="Square"):
        super().__init__(id, name)

    def draw(self):
        print("Draw a Square")

    def fill(self, color):
        print(f"Fill with {Color.to_string(color)}")

    def clone(self):
        return Square(self._id, self._name)

class Rectangle(Shape):
    def __init__(self, id=ShapeKind.Rectangle, name="Rectangle"):
        super().__init__(id, name)

    def draw(self):
        print("Draw a Rectangle")

    def fill(self, color):
        print(f"Fill with {Color.to_string(color)}")

    def clone(self):
        return Rectangle(self._id, self._name)

File: 01_GOF23/DP_examples/Lua/01_Creational/Prototype/Product.lua
-- 颜色枚举
---@enum Color
Color = {
    Unknown = 0,
    Red = 1,
    Green = 2,
    Blue = 3,
    Yellow = 4
}

-- 颜色转换为字符串
---@param color Color
---@return string
local function ColorToString(color)
    local colorMap = {
        [Color.Red] = "Red",
        [Color.Green] = "Green",
        [Color.Blue] = "Blue",
        [Color.Yellow] = "Yellow"
    }
    return colorMap[color] or "Unknown"
end

-- 形状接口
IShape = {
    id = 0,
    name = ""
}
function IShape:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

function IShape:name()
    return self.name
end

function IShape:draw() end

function IShape:fill(color) end

function IShape:clone() end

-- 形状基类
---@class Shape : IShape
Shape = IShape:new()
function Shape:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

-- 形状类型枚举
ShapeKind = {
    Unknown = 0,
    Circle = 1,
    Square = 2,
    Rectangle = 3
}

-- 圆形
---@class Circle : Shape
Circle = Shape:new()
function Circle:new(o)
    o = o or {}
    o.id = o.id or ShapeKind.Circle
    o.name = o.name or "Circle"
    setmetatable(o, self)
    self.__index = self
    return o
end

function Circle:draw()
    print("Draw a Circle")
end

function Circle:fill(color)
    print("Fill with " .. ColorToString(color))
end

function Circle:clone()
    return Circle:new({ id = self.id, name = self.name })
end

-- 正方形
---@class Square : Shape
Square = Shape:new()
function Square:new(o)
    o = o or {}
    o.id = o.id or ShapeKind.Square
    o.name = o.name or "Square"
    setmetatable(o, self)
    self.__index = self
    return o
end

function Square:draw()
    print("Draw a Square")
end

function Square:fill(color)
    print("Fill with " .. ColorToString(color))
end

function Square:clone()
    return Square:new({ id = self.id, name = self.name })
end

-- 矩形
---@class Rectangle : Shape
Rectangle = Shape:new()
function Rectangle:new(o)
    o = o or {}
    o.id = o.id or ShapeKind.Rectangle
    o.name = o.name or "Rectangle"
    setmetatable(o, self)
    self.__index = self
    return o
end

function Rectangle:draw()
    print("Draw a Rectangle")
end

function Rectangle:fill(color)
    print("Fill with " .. ColorToString(color))
end

function Rectangle:clone()
    return Rectangle:new({ id = self.id, name = self.name })
end
